tests/integration/olp-cpp-sdk-dataservice-write/VersionedLayerClientTest.cpp

```diff
@@ -0,0 +1,310 @@
+
+class VersionedLayerClientTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    network_ = std::make_shared<NetworkMock>();
+    settings_.cache =
+        olp::client::OlpClientSettingsFactory::CreateDefaultCache({});
+    settings_.network_request_handler = network_;
+    settings_.task_scheduler =
+        olp::client::OlpClientSettingsFactory::CreateDefaultTaskScheduler(1);
+  }
+
+  void TearDown() override {
+    settings_.network_request_handler.reset();
+    settings_.cache.reset();
+
+    network_.reset();
+  }
+
+  write::model::Apis CreateApiResponse(const std::string& service) {
+    const auto apis =
+        mockserver::DefaultResponses::GenerateResourceApisResponse(
+            kHrn.ToCatalogHRNString());
+    auto it = std::find_if(
+        apis.begin(), apis.end(),
+        [&](const model::Api& obj) -> bool { return obj.GetApi() == service; });
+
+    write::model::Apis result;
+    if (it != apis.end()) {
+      result.push_back(*it);
+    }
+
+    return result;
+  }
+
+  std::shared_ptr<NetworkMock> network_;
+  olp::client::OlpClientSettings settings_;
+};
+
+TEST_F(VersionedLayerClientTest, StartBatch) {
+  const auto catalog = kHrn.ToCatalogHRNString();
+  const auto api = CreateApiResponse(kPublishApiName);
+  const auto publication =
+      mockserver::DefaultResponses::GeneratePublicationResponse({kLayer}, {});
+  ASSERT_FALSE(api.empty());
+
+  {
+    SCOPED_TRACE("Success request");
+
+    EXPECT_CALL(*network_, Send(IsGetRequest(kLookupPublishApiUrl), _, _, _, _))
+        .WillOnce(ReturnHttpResponse(olp::http::NetworkResponse().WithStatus(
+                                         olp::http::HttpStatusCode::OK),
+                                     olp::serializer::serialize(api)));
+    EXPECT_CALL(*network_, Send(IsPostRequest(kPublishUrl), _, _, _, _))
+        .WillOnce(ReturnHttpResponse(olp::http::NetworkResponse().WithStatus(
+                                         olp::http::HttpStatusCode::OK),
+                                     olp::serializer::serialize(publication)));
+
+    write::VersionedLayerClient write_client(kHrn, settings_);
+    const auto batch_request = model::StartBatchRequest().WithLayers({kLayer});
+    const auto response =
+        write_client.StartBatch(batch_request).GetFuture().get();
+    const auto& result = response.GetResult();
+
+    EXPECT_TRUE(response.IsSuccessful());
+    ASSERT_TRUE(result.GetId());
+    ASSERT_TRUE(result.GetDetails());
+    ASSERT_TRUE(result.GetLayerIds());
+    ASSERT_EQ(result.GetLayerIds().get().size(), 1);
+    ASSERT_EQ(result.GetLayerIds().get().front(), kLayer);
+    ASSERT_NE("", response.GetResult().GetId().value());
+    Mock::VerifyAndClearExpectations(network_.get());
+  }
+
+  {
+    SCOPED_TRACE("Success request, but with callback");
+
+    // don't mock lookup api request since its cached in previous step
+    EXPECT_CALL(*network_, Send(IsPostRequest(kPublishUrl), _, _, _, _))
```

Comment:
Ideally, you would test here that it takes the lookup API from the cache and not requesting it again.

Reply:
Test will fail if lookup API request occur.
Anyway, added mock with `Times(0)` so its clear API requests are not welcome here :-)